src/storage/mod.rs

```diff
@@ -155,7 +156,7 @@ pub type Callback<T> = Box<dyn FnOnce(Result<T>) + Send>;
 /// that multiple versions can be saved at the same time. Raw operations use raw
 /// keys, which are saved directly to the engine without memcomparable- encoding
 /// and appending timestamp.
-pub struct Storage<E: Engine, L: LockManager, F: KvFormat> {
+pub struct Storage<E: Engine, L: LockManager, F: KvFormat, Ts: CausalTsProvider + 'static> {
```

Comment:
The `CausalTsProvider` trait is spread all over the code, and only introduced for test. I think it's unworthy and affect compile time. Not sure whether the dynamic dispatch would make really influence here. 

Maybe we can just write the test in another way to avoid using trait or use(enum_dispatch)[https://docs.rs/enum_dispatch/latest/enum_dispatch/]. This has been a selection problem troubling me for a long time. What's the best practice? /cc @BusyJay

Reply:
I see the problem. I think there are three solutions here:
1. Use `dyn` type, e.g. `Option<Arc<dyn CausalTsProvider>>` in `Storage`, the drawback is may consuem more cpu for vtable lookup, considerding it's in critiical io path, I think it's not the best practice.
2. Use concrete type, e.g. `Option<Arc<BatchProvider<RpcClient>>>`, the problem here is bring the dependence of `RpcClient` for `Storage`, which break the possibility of `Storage` actually can be a stand-alone storage.
3. Use `enum_dispatch`.

I would like to try the third solution. How do you think? @pingyu @BusyJay

Reply:
`enum_dispatch` +1

Reply:
If there will be only one production provider, you can make it the default.
```
pub trait Storage<TsProvider = BatchProvider<RpcClient>> {
```

Reply:
Thanks for Jay's great suggestion. After a few attempts, I found that a few unit test cases are creating `Storage` or `tikv::Server` or `tikv::server::Service`, which all need to specify the trait type of `Storage`, e.g. [`test_peer_resolve`](https://github.com/tikv/tikv/blob/4f717716a515a8047d91f4cbeef219f41c20dc2c/src/server/server.rs#L482-L560). This bring much code changes. After discussion, we decide to use `enum_dispatch`. Thanks for you all.